### PR TITLE
[Plugins] Include Plugins in `Manifest.targetsRequired`

### PR DESCRIPTION
**Pull request by fourplusone**
Include Plugins in `Manifest.targetsRequired`

### Motivation:

Fixes SR-14343. As Plugins are modeled as a dependency they also must be included the list of required targets for a product.

### Modifications:

Changed the implementation of `Manifest.targetsRequired` to include plugins 

### Result:

SR-14343 is fixed

### CODE REVIEW

**Review by neonichu — APPROVED** [review]

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: Sources/PackageModel/Manifest.swift [hunk 1]
```diff
@@ -202,15 +202,33 @@ public final class Manifest: ObjectIdentifierProtocol {
...
+                        return targetsByName.keys.contains(name) ? name : nil
+                    default:
+                        return nil
+                    }
                 }
-            }) ?? []
+
+                let plugins: [String] = target.pluginUsages?.compactMap { pluginUsage in
+                    switch pluginUsage {
+                    case .plugin(name: let name, package: nil):
```
Comment (tomerd) [inline Sources/PackageModel/Manifest.swift:219]: should this be `.plugin(name: let name,  _):`?

Comment (fourplusone) [reply Sources/PackageModel/Manifest.swift:219]: My understanding ist, that if package is not `nil` `.plugin(...)` declares a dependency to a target in another package. Therefore it's not relevant in this context

**Comment by neonichu** [comment]
> 10:23:08 Test Suite 'PackageCollectionsTests' started at 2021-07-20 10:21:53.935
> 10:23:08 Test Case '-[PackageCollectionsTests.PackageCollectionsTests testCollectionPendingTrustConfirmIsKeptOnAdd]' started.
> 10:23:08 Test Case '-[PackageCollectionsTests.PackageCollectionsTests testCollectionPendingTrustConfirmIsKeptOnAdd]' passed (0.105 seconds).
> 10:23:08 Exited with signal code 4

Maybe interesting for @yim-lee

**Comment by neonichu** [comment]
@swift-ci please smoke test macOS

**Comment by yim-lee** [comment]
https://ci.swift.org/job/swift-package-manager-with-xcode-self-hosted-PR-osx/2239/console

Hard to tell what happened with `PackageCollectionsTests`. Will keep an eye on it in case it happens again.

**Comment by abertelrud** [comment]
Thanks for the fix, @fourplusone!  The fix looks good... it would be great to also have a unit test for it.  There are some test fixtures for plugins in the _Fixtures/Miscellaneous/Plugins_ subdirectory that could be useful for this purpose.

**Review by abertelrud — APPROVED** [review]

**Comment by elsh** [comment]
> Thanks for the fix, @fourplusone! The fix looks good... it would be great to also have a unit test for it. There are some test fixtures for plugins in the _Fixtures/Miscellaneous/Plugins_ subdirectory that could be useful for this purpose.

@fourplusone do you need any additional info on adding unit tests?

**Comment by abertelrud** [comment]
Extended a unit test to cover this.  It's at https://github.com/abertelrud/swift-package-manager/commit/c7fd9ece77ea0ea617bdcc0f15206029b76de433.  @fourplusone, do you mind if I add that to this branch?

**Comment by fourplusone** [comment]
@abertelrud Feel free to add it 🙂

**Comment by abertelrud** [comment]
> @abertelrud Feel free to add it 🙂

Great, thanks!  I pushed a commit extending an existing test to cover this case.  Thank you for the fix!

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test linux

**Comment by abertelrud** [comment]
Looks like all the tests pass and the reviews are in place, so merging this.  Thank you for the fix!

**Pull request merged**